{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ffec50e7-f926-4d36-bcfc-a04284d57c33",
   "metadata": {},
   "source": [
    "## Overview\n",
    "Microarray expression data is measured at the probe level, where multiple probes may map to the same gene or transcript.\n",
    "In this notebook, we:\n",
    "1. Load the Affymetrix GPL570 platform annotation\n",
    "2. Map probe-level differential expression results to genes using Entrez Gene IDs\n",
    "3. Resolve multiple probes per gene using a principled aggregation strategy\n",
    "4. Generate a gene-level differential expression table suitable for downstream functional and pathway analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2c822d9c-687a-4d60-ad1a-fa3390bfd295",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f27b8d50-b3ac-4316-822e-09098b7f9da6",
   "metadata": {},
   "source": [
    "#### Probe-to-Gene Mapping and Systems-Level Interpretation\r\n",
    "Differential expression analysis in the previous notebook identified probe-level transcriptional changes between Duchenne muscular dystrophy (DMD) patient muscle samples and controls. However, Affymetrix microarray platforms measure expression at the probe level, and multiple probes may map to the same gene. To enable biologically meaningful interpretation and downstream pathway analysis, differentially expressed probes must therefore be mapped to gene-level identifiers and consolidated appropriately. This notebook focuses on translating probe-level differential expression results into gene-level signals and interpreting these changes in the context of molecular pathways and disease mechanisms relevant to DMD\n",
    "\r\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "014695c8-aec8-4d8d-814c-f541bc1fb1f1",
   "metadata": {},
   "source": [
    "##### Probe Annotation Strategy\r\n",
    "Affymetrix probes can map to one or more genes, and a single gene may be represented by multiple probes. To ensure biologically interpretable results, probes will be mapped to gene symbols using platform-specific annotation resources. When multiple probes map to the same gene, probe-level signals will be consolidated to generate a single gene-level differential expression value per gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "152c0f57-8116-45b7-9d02-f683a0df7057",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mean_control</th>\n",
       "      <th>mean_dmd</th>\n",
       "      <th>log2FC</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>probe_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>34471_at</th>\n",
       "      <td>4.320852</td>\n",
       "      <td>13.744169</td>\n",
       "      <td>9.423317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205940_at</th>\n",
       "      <td>5.229859</td>\n",
       "      <td>13.853884</td>\n",
       "      <td>8.624025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206717_at</th>\n",
       "      <td>5.431540</td>\n",
       "      <td>13.735486</td>\n",
       "      <td>8.303946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202311_s_at</th>\n",
       "      <td>4.035683</td>\n",
       "      <td>10.616064</td>\n",
       "      <td>6.580381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209875_s_at</th>\n",
       "      <td>2.515965</td>\n",
       "      <td>9.046366</td>\n",
       "      <td>6.530402</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             mean_control   mean_dmd    log2FC\n",
       "probe_id                                      \n",
       "34471_at         4.320852  13.744169  9.423317\n",
       "205940_at        5.229859  13.853884  8.624025\n",
       "206717_at        5.431540  13.735486  8.303946\n",
       "202311_s_at      4.035683  10.616064  6.580381\n",
       "209875_s_at      2.515965   9.046366  6.530402"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load probe-level DE results from Notebook 2\n",
    "de_dmd_ctrl = pd.read_csv(\n",
    "    r\"C:\\Users\\Lenovo\\Desktop\\UG Stream\\Notebooks\\02 - DE_DMD_vs_Control.csv\",\n",
    "    index_col=0\n",
    ")\n",
    "\n",
    "de_dmd_ctrl.index.name = \"probe_id\"\n",
    "\n",
    "de_dmd_ctrl.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "793cb138-4b41-4d6a-8405-780f16e57eca",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert de_dmd_ctrl.index.str.contains(\"_at\").any()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "cf62d842-90fe-44e2-bbef-1f95f8077a86",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>GB_ACC</th>\n",
       "      <th>SPOT_ID</th>\n",
       "      <th>Species Scientific Name</th>\n",
       "      <th>Annotation Date</th>\n",
       "      <th>Sequence Type</th>\n",
       "      <th>Sequence Source</th>\n",
       "      <th>Target Description</th>\n",
       "      <th>Representative Public ID</th>\n",
       "      <th>Gene Title</th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>ENTREZ_GENE_ID</th>\n",
       "      <th>RefSeq Transcript ID</th>\n",
       "      <th>Gene Ontology Biological Process</th>\n",
       "      <th>Gene Ontology Cellular Component</th>\n",
       "      <th>Gene Ontology Molecular Function</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1007_s_at</td>\n",
       "      <td>U48705</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>Affymetrix Proprietary Database</td>\n",
       "      <td>U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Huma...</td>\n",
       "      <td>U48705</td>\n",
       "      <td>discoidin domain receptor tyrosine kinase 1 //...</td>\n",
       "      <td>DDR1 /// MIR4640</td>\n",
       "      <td>780 /// 100616237</td>\n",
       "      <td>NM_001202521 /// NM_001202522 /// NM_001202523...</td>\n",
       "      <td>0001558 // regulation of cell growth // inferr...</td>\n",
       "      <td>0005576 // extracellular region // inferred fr...</td>\n",
       "      <td>0000166 // nucleotide binding // inferred from...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1053_at</td>\n",
       "      <td>M87338</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>M87338 /FEATURE= /DEFINITION=HUMA1SBU Human re...</td>\n",
       "      <td>M87338</td>\n",
       "      <td>replication factor C (activator 1) 2, 40kDa</td>\n",
       "      <td>RFC2</td>\n",
       "      <td>5982</td>\n",
       "      <td>NM_001278791 /// NM_001278792 /// NM_001278793...</td>\n",
       "      <td>0000278 // mitotic cell cycle // traceable aut...</td>\n",
       "      <td>0005634 // nucleus // inferred from electronic...</td>\n",
       "      <td>0000166 // nucleotide binding // inferred from...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>117_at</td>\n",
       "      <td>X51757</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>Affymetrix Proprietary Database</td>\n",
       "      <td>X51757 /FEATURE=cds /DEFINITION=HSP70B Human h...</td>\n",
       "      <td>X51757</td>\n",
       "      <td>heat shock 70kDa protein 6 (HSP70B')</td>\n",
       "      <td>HSPA6</td>\n",
       "      <td>3310</td>\n",
       "      <td>NM_002155</td>\n",
       "      <td>0000902 // cell morphogenesis // inferred from...</td>\n",
       "      <td>0005737 // cytoplasm // inferred from direct a...</td>\n",
       "      <td>0000166 // nucleotide binding // inferred from...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>121_at</td>\n",
       "      <td>X69699</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens...</td>\n",
       "      <td>X69699</td>\n",
       "      <td>paired box 8</td>\n",
       "      <td>PAX8</td>\n",
       "      <td>7849</td>\n",
       "      <td>NM_003466 /// NM_013951 /// NM_013952 /// NM_0...</td>\n",
       "      <td>0001655 // urogenital system development // in...</td>\n",
       "      <td>0005634 // nucleus // inferred from direct ass...</td>\n",
       "      <td>0000979 // RNA polymerase II core promoter seq...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1255_g_at</td>\n",
       "      <td>L36861</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Oct 6, 2014</td>\n",
       "      <td>Exemplar sequence</td>\n",
       "      <td>Affymetrix Proprietary Database</td>\n",
       "      <td>L36861 /FEATURE=expanded_cds /DEFINITION=HUMGC...</td>\n",
       "      <td>L36861</td>\n",
       "      <td>guanylate cyclase activator 1A (retina)</td>\n",
       "      <td>GUCA1A</td>\n",
       "      <td>2978</td>\n",
       "      <td>NM_000409 /// XM_006715073</td>\n",
       "      <td>0007165 // signal transduction // non-traceabl...</td>\n",
       "      <td>0001750 // photoreceptor outer segment // infe...</td>\n",
       "      <td>0005509 // calcium ion binding // inferred fro...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          ID  GB_ACC SPOT_ID Species Scientific Name Annotation Date  \\\n",
       "0  1007_s_at  U48705     NaN            Homo sapiens     Oct 6, 2014   \n",
       "1    1053_at  M87338     NaN            Homo sapiens     Oct 6, 2014   \n",
       "2     117_at  X51757     NaN            Homo sapiens     Oct 6, 2014   \n",
       "3     121_at  X69699     NaN            Homo sapiens     Oct 6, 2014   \n",
       "4  1255_g_at  L36861     NaN            Homo sapiens     Oct 6, 2014   \n",
       "\n",
       "       Sequence Type                  Sequence Source  \\\n",
       "0  Exemplar sequence  Affymetrix Proprietary Database   \n",
       "1  Exemplar sequence                          GenBank   \n",
       "2  Exemplar sequence  Affymetrix Proprietary Database   \n",
       "3  Exemplar sequence                          GenBank   \n",
       "4  Exemplar sequence  Affymetrix Proprietary Database   \n",
       "\n",
       "                                  Target Description Representative Public ID  \\\n",
       "0  U48705 /FEATURE=mRNA /DEFINITION=HSU48705 Huma...                   U48705   \n",
       "1  M87338 /FEATURE= /DEFINITION=HUMA1SBU Human re...                   M87338   \n",
       "2  X51757 /FEATURE=cds /DEFINITION=HSP70B Human h...                   X51757   \n",
       "3  X69699 /FEATURE= /DEFINITION=HSPAX8A H.sapiens...                   X69699   \n",
       "4  L36861 /FEATURE=expanded_cds /DEFINITION=HUMGC...                   L36861   \n",
       "\n",
       "                                          Gene Title       Gene Symbol  \\\n",
       "0  discoidin domain receptor tyrosine kinase 1 //...  DDR1 /// MIR4640   \n",
       "1        replication factor C (activator 1) 2, 40kDa              RFC2   \n",
       "2               heat shock 70kDa protein 6 (HSP70B')             HSPA6   \n",
       "3                                       paired box 8              PAX8   \n",
       "4            guanylate cyclase activator 1A (retina)            GUCA1A   \n",
       "\n",
       "      ENTREZ_GENE_ID                               RefSeq Transcript ID  \\\n",
       "0  780 /// 100616237  NM_001202521 /// NM_001202522 /// NM_001202523...   \n",
       "1               5982  NM_001278791 /// NM_001278792 /// NM_001278793...   \n",
       "2               3310                                          NM_002155   \n",
       "3               7849  NM_003466 /// NM_013951 /// NM_013952 /// NM_0...   \n",
       "4               2978                         NM_000409 /// XM_006715073   \n",
       "\n",
       "                    Gene Ontology Biological Process  \\\n",
       "0  0001558 // regulation of cell growth // inferr...   \n",
       "1  0000278 // mitotic cell cycle // traceable aut...   \n",
       "2  0000902 // cell morphogenesis // inferred from...   \n",
       "3  0001655 // urogenital system development // in...   \n",
       "4  0007165 // signal transduction // non-traceabl...   \n",
       "\n",
       "                    Gene Ontology Cellular Component  \\\n",
       "0  0005576 // extracellular region // inferred fr...   \n",
       "1  0005634 // nucleus // inferred from electronic...   \n",
       "2  0005737 // cytoplasm // inferred from direct a...   \n",
       "3  0005634 // nucleus // inferred from direct ass...   \n",
       "4  0001750 // photoreceptor outer segment // infe...   \n",
       "\n",
       "                    Gene Ontology Molecular Function  \n",
       "0  0000166 // nucleotide binding // inferred from...  \n",
       "1  0000166 // nucleotide binding // inferred from...  \n",
       "2  0000166 // nucleotide binding // inferred from...  \n",
       "3  0000979 // RNA polymerase II core promoter seq...  \n",
       "4  0005509 // calcium ion binding // inferred fro...  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot_df = pd.read_csv(\n",
    "    r\"C:\\Users\\Lenovo\\Desktop\\UG Stream\\GPL570-55999 (1).txt\",\n",
    "    sep=\"\\t\",\n",
    "    comment=\"#\",\n",
    "    low_memory=False\n",
    ")\n",
    "\n",
    "annot_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "04bd6070-166f-46b7-b2d1-4e3af20a8f7c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>probe_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>entrez_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1007_s_at</td>\n",
       "      <td>DDR1 /// MIR4640</td>\n",
       "      <td>780 /// 100616237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1053_at</td>\n",
       "      <td>RFC2</td>\n",
       "      <td>5982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>117_at</td>\n",
       "      <td>HSPA6</td>\n",
       "      <td>3310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>121_at</td>\n",
       "      <td>PAX8</td>\n",
       "      <td>7849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1255_g_at</td>\n",
       "      <td>GUCA1A</td>\n",
       "      <td>2978</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    probe_id       gene_symbol          entrez_id\n",
       "0  1007_s_at  DDR1 /// MIR4640  780 /// 100616237\n",
       "1    1053_at              RFC2               5982\n",
       "2     117_at             HSPA6               3310\n",
       "3     121_at              PAX8               7849\n",
       "4  1255_g_at            GUCA1A               2978"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot_df = annot_df[[\n",
    "    \"ID\",\n",
    "    \"Gene Symbol\",\n",
    "    \"ENTREZ_GENE_ID\"\n",
    "]].rename(columns={\n",
    "    \"ID\": \"probe_id\",\n",
    "    \"Gene Symbol\": \"gene_symbol\",\n",
    "    \"ENTREZ_GENE_ID\": \"entrez_id\"\n",
    "})\n",
    "\n",
    "annot_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "fde583f6-8b99-48c2-a38f-2efeb2dd0667",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(54675, 3)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "dc2f1fb6-e770-4ff5-b750-0e879bea7a82",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot_df[\"probe_id\"].isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "975c9599-b57c-4478-ad59-3fdbe403cafe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10541"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot_df[\"entrez_id\"].isna().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c88f0e5c-c888-4fbd-b0c8-9764160bba7d",
   "metadata": {},
   "source": [
    "Not all probes map to an Entrez Gene ID. Probes without gene assignments are excluded from gene-level analyses."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "dbf05038-2038-4a49-99d1-c91d94660a79",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clean Entrez IDs: keep primary mapping if multiple IDs exist\n",
    "annot_df[\"entrez_id\"] = (\n",
    "    annot_df[\"entrez_id\"]\n",
    "    .astype(str)\n",
    "    .str.split(\" /// \")\n",
    "    .str[0]\n",
    ")\n",
    "\n",
    "# Convert to numeric (invalid entries become NaN)\n",
    "annot_df[\"entrez_id\"] = pd.to_numeric(annot_df[\"entrez_id\"], errors=\"coerce\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "a94776ba-8eca-45dd-a49a-d8180affec57",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      780.0\n",
       "1     5982.0\n",
       "2     3310.0\n",
       "3     7849.0\n",
       "4     2978.0\n",
       "5     7318.0\n",
       "6     7067.0\n",
       "7    11099.0\n",
       "8     6352.0\n",
       "9     1571.0\n",
       "Name: entrez_id, dtype: float64"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot_df[\"entrez_id\"].head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "d7f85506-8991-413f-ab78-2908e197e4cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10541"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot_df[\"entrez_id\"].isna().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87b02aa7-7fc8-4506-b9dc-c5bad9df76de",
   "metadata": {},
   "source": [
    "## Handling multiple gene mappings\n",
    "Some Affymetrix probes map to multiple genes, represented as multiple Entrez Gene IDs separated by /// in the platform annotation.\n",
    "For downstream gene-level analysis, only the primary Entrez Gene ID (first listed) was retained for each probe. Probes without Entrez Gene ID assignments were excluded from gene-level aggregation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "105f869d-80fc-4703-aa99-76a4b64d3b3f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mean_control</th>\n",
       "      <th>mean_dmd</th>\n",
       "      <th>log2FC</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>probe_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>34471_at</th>\n",
       "      <td>4.320852</td>\n",
       "      <td>13.744169</td>\n",
       "      <td>9.423317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205940_at</th>\n",
       "      <td>5.229859</td>\n",
       "      <td>13.853884</td>\n",
       "      <td>8.624025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206717_at</th>\n",
       "      <td>5.431540</td>\n",
       "      <td>13.735486</td>\n",
       "      <td>8.303946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202311_s_at</th>\n",
       "      <td>4.035683</td>\n",
       "      <td>10.616064</td>\n",
       "      <td>6.580381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209875_s_at</th>\n",
       "      <td>2.515965</td>\n",
       "      <td>9.046366</td>\n",
       "      <td>6.530402</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             mean_control   mean_dmd    log2FC\n",
       "probe_id                                      \n",
       "34471_at         4.320852  13.744169  9.423317\n",
       "205940_at        5.229859  13.853884  8.624025\n",
       "206717_at        5.431540  13.735486  8.303946\n",
       "202311_s_at      4.035683  10.616064  6.580381\n",
       "209875_s_at      2.515965   9.046366  6.530402"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "de_dmd_ctrl = de_dmd_ctrl.copy()\n",
    "de_dmd_ctrl.index.name = \"probe_id\"\n",
    "de_dmd_ctrl.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "92f8ce2e-8166-4954-ad22-ab73fc9840b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert de_dmd_ctrl.index.str.contains(\"_at\").any()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "4a2d06f0-4b3e-4705-ad9d-a91619ecc09c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>probe_id</th>\n",
       "      <th>mean_control</th>\n",
       "      <th>mean_dmd</th>\n",
       "      <th>log2FC</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>entrez_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>34471_at</td>\n",
       "      <td>4.320852</td>\n",
       "      <td>13.744169</td>\n",
       "      <td>9.423317</td>\n",
       "      <td>MYH8</td>\n",
       "      <td>4626.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>205940_at</td>\n",
       "      <td>5.229859</td>\n",
       "      <td>13.853884</td>\n",
       "      <td>8.624025</td>\n",
       "      <td>MYH3</td>\n",
       "      <td>4621.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>206717_at</td>\n",
       "      <td>5.431540</td>\n",
       "      <td>13.735486</td>\n",
       "      <td>8.303946</td>\n",
       "      <td>MYH8</td>\n",
       "      <td>4626.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>202311_s_at</td>\n",
       "      <td>4.035683</td>\n",
       "      <td>10.616064</td>\n",
       "      <td>6.580381</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>1277.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>209875_s_at</td>\n",
       "      <td>2.515965</td>\n",
       "      <td>9.046366</td>\n",
       "      <td>6.530402</td>\n",
       "      <td>SPP1</td>\n",
       "      <td>6696.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      probe_id  mean_control   mean_dmd    log2FC gene_symbol  entrez_id\n",
       "0     34471_at      4.320852  13.744169  9.423317        MYH8     4626.0\n",
       "1    205940_at      5.229859  13.853884  8.624025        MYH3     4621.0\n",
       "2    206717_at      5.431540  13.735486  8.303946        MYH8     4626.0\n",
       "3  202311_s_at      4.035683  10.616064  6.580381      COL1A1     1277.0\n",
       "4  209875_s_at      2.515965   9.046366  6.530402        SPP1     6696.0"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "de_annot = (\n",
    "    de_dmd_ctrl\n",
    "    .reset_index()\n",
    "    .merge(\n",
    "        annot_df,\n",
    "        on=\"probe_id\",\n",
    "        how=\"left\"\n",
    "    )\n",
    ")\n",
    "\n",
    "de_annot.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "d7747aac-76bf-4c54-a486-6eda8a9b3e07",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DE rows (probes): 54675\n",
      "Merged rows: 54675\n"
     ]
    }
   ],
   "source": [
    "# Shape check\n",
    "print(\"DE rows (probes):\", de_dmd_ctrl.shape[0])\n",
    "print(\"Merged rows:\", de_annot.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "1a760f09-aed1-4ec4-8598-2886e3635c0a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "44134"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# How many probes got an Entrez ID?\n",
    "de_annot[\"entrez_id\"].notna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "bb3be541-6d62-4e1a-b7e0-1565d63d07f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "54675"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# How many unique probes survived?\n",
    "de_annot[\"probe_id\"].nunique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cb117d1-e5cb-491b-be2d-f19a783d0a9f",
   "metadata": {},
   "source": [
    "### Probe-Level Annotation\r\n",
    "Probe-level differential expression results were merged with the Affymetrix GPL570 platform annotation using probe identifiers (ID_REF). This step associates each probe with its corresponding gene symbol and Entrez Gene ID where available. Probes without gene assignments were retained at this stage and excluded only during gene-level aggregation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "70f3ab15-567d-4f3e-a9ed-7b39259e5a3e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Probes before filtering: 54675\n",
      "Probes after filtering: 44134\n"
     ]
    }
   ],
   "source": [
    "# Keep only probes with valid Entrez Gene IDs\n",
    "de_annot_valid = de_annot.dropna(subset=[\"entrez_id\"]).copy()\n",
    "\n",
    "print(\"Probes before filtering:\", de_annot.shape[0])\n",
    "print(\"Probes after filtering:\", de_annot_valid.shape[0])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "40e72baf-5a46-444b-9f21-d9a222c23a52",
   "metadata": {},
   "outputs": [],
   "source": [
    "de_annot_valid[\"entrez_id\"] = de_annot_valid[\"entrez_id\"].astype(\"Int64\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d03b7b71-bfe5-456b-86c8-a34156285842",
   "metadata": {},
   "source": [
    "#### Filtering probes without gene assignments\n",
    "Not all Affymetrix probes map to a known gene. Probes lacking Entrez Gene ID annotations were excluded prior to gene-level aggregation to ensure that downstream analyses operate on biologically interpretable units."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "8a89835e-f213-41ea-8c43-40379de2a56e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mean_log2FC</th>\n",
       "      <th>n_probes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>entrez_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4626</th>\n",
       "      <td>8.863631</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4621</th>\n",
       "      <td>8.624025</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5320</th>\n",
       "      <td>6.528468</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4636</th>\n",
       "      <td>6.368076</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158763</th>\n",
       "      <td>5.804189</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           mean_log2FC  n_probes\n",
       "entrez_id                       \n",
       "4626          8.863631         2\n",
       "4621          8.624025         1\n",
       "5320          6.528468         1\n",
       "4636          6.368076         1\n",
       "158763        5.804189         1"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_de = (\n",
    "    de_annot_valid\n",
    "    .groupby(\"entrez_id\")\n",
    "    .agg(\n",
    "        mean_log2FC=(\"log2FC\", \"mean\"),\n",
    "        n_probes=(\"probe_id\", \"count\")\n",
    "    )\n",
    "    .sort_values(\"mean_log2FC\", ascending=False)\n",
    ")\n",
    "\n",
    "gene_de.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "dfe914e8-b883-435e-920f-e4112db8b5cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(21180, 2)"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_de.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "952ebe07-79fc-404e-93d8-1071fed7f6bd",
   "metadata": {},
   "source": [
    "#### Gene-level aggregation strategy\n",
    "\r\n",
    "Multiple Affymetrix probes may target the same gene due to alternative transcripts or probe redundanc.\r\n",
    "\r\n",
    "To obtain a single gene-level estimate of differential expression, probe-level log2 fold changes were aggregated by mean log2 fold change per Entrez GeneID.\r\n",
    "\r\n",
    "The number of contributing probes per gene was retained to provide transparency regarding probe coverage."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "baf2d29f-a711-4557-89a4-34a8fc656023",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mean_log2FC</th>\n",
       "      <th>n_probes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>entrez_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4626</th>\n",
       "      <td>8.863631</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4621</th>\n",
       "      <td>8.624025</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5320</th>\n",
       "      <td>6.528468</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4636</th>\n",
       "      <td>6.368076</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158763</th>\n",
       "      <td>5.804189</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5816</th>\n",
       "      <td>5.747369</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4969</th>\n",
       "      <td>5.123637</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6192</th>\n",
       "      <td>4.888431</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100506621</th>\n",
       "      <td>4.655324</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55228</th>\n",
       "      <td>4.523857</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           mean_log2FC  n_probes\n",
       "entrez_id                       \n",
       "4626          8.863631         2\n",
       "4621          8.624025         1\n",
       "5320          6.528468         1\n",
       "4636          6.368076         1\n",
       "158763        5.804189         1\n",
       "5816          5.747369         1\n",
       "4969          5.123637         2\n",
       "6192          4.888431         1\n",
       "100506621     4.655324         1\n",
       "55228         4.523857         1"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Most upregulated genes\n",
    "gene_de.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "d46a626e-0567-42be-b8e9-44776ef97008",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mean_log2FC</th>\n",
       "      <th>n_probes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>entrez_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>165186</th>\n",
       "      <td>-4.014346</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10647</th>\n",
       "      <td>-4.183059</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10265</th>\n",
       "      <td>-4.218394</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8549</th>\n",
       "      <td>-4.231001</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340156</th>\n",
       "      <td>-4.303520</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26628</th>\n",
       "      <td>-4.351001</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9211</th>\n",
       "      <td>-4.485525</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115019</th>\n",
       "      <td>-4.507204</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64850</th>\n",
       "      <td>-4.524759</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7306</th>\n",
       "      <td>-5.708823</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           mean_log2FC  n_probes\n",
       "entrez_id                       \n",
       "165186       -4.014346         1\n",
       "10647        -4.183059         1\n",
       "10265        -4.218394         1\n",
       "8549         -4.231001         2\n",
       "340156       -4.303520         3\n",
       "26628        -4.351001         1\n",
       "9211         -4.485525         1\n",
       "115019       -4.507204         1\n",
       "64850        -4.524759         1\n",
       "7306         -5.708823         1"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Most downregulated genes\n",
    "gene_de.tail(10)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "a7eb8ff9-eddf-4d19-aeb5-629b8d817df8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    21180.000000\n",
       "mean         0.052062\n",
       "std          0.648572\n",
       "min         -5.708823\n",
       "25%         -0.173561\n",
       "50%         -0.000061\n",
       "75%          0.192173\n",
       "max          8.863631\n",
       "Name: mean_log2FC, dtype: float64"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_de[\"mean_log2FC\"].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "b215c763-4e93-49c0-8146-d834b968ef02",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved\n"
     ]
    }
   ],
   "source": [
    "gene_de.to_csv(\n",
    "    r\"C:\\Users\\Lenovo\\Desktop\\UG Stream\\Notebooks\\03 - Gene_Level_DE_Entrez.csv\"\n",
    ")\n",
    "annot_df.to_csv(\n",
    "    r\"C:\\Users\\Lenovo\\Desktop\\UG Stream\\Notebooks\\03 - Probe_to_Gene_Annotation.csv\",\n",
    "    index=False\n",
    ")\n",
    "print(\"Saved\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60a6d85c-d650-4eef-bbea-197eb37d4900",
   "metadata": {},
   "source": [
    "### Notebook 03 - Summary and Next Steps\r\n",
    "\r\n",
    "In this notebook, we transitioned from probe-level differential expression results to a biologically interpretable **gene-level differential expression profile** for Duchenne Muscular Dystrophy (DMD).\r\n",
    "\r\n",
    "Specifically, we:\r\n",
    "- Loaded the Affymetrix GPL570 platform annotation and extracted probe-to-gene mappings  \r\n",
    "- Resolved probes mapping to multiple genes by retaining the primary Entrez Gene ID  \r\n",
    "- Merged probe-level differential expression results with annotation using probe identifiers  \r\n",
    "- Excluded probes without valid Entrez Gene ID assignments  \r\n",
    "- Aggregated probe-level log2 fold changes to obtain **gene-level differential expression estimates**, while retaining probe counts per gene  \r\n",
    "\r\n",
    "The final output of this notebook is a curated **gene-level differential expression table**, indexed by Entrez Gene ID, suitable for downstream functional analysis and target prioritization.\r\n",
    "\r\n",
    "**Next steps** will involve leveraging this gene-level profile to perform higher-level biological interpretation, including pathway and functional enrichment analyses, disease stageâ€“specific expression comparisons, and prioritization of candidate genes for therapeutic targeting and molecular docking studies.\r\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "149ffd72-0ccf-4071-a5ef-91e70a1ad6a8",
   "metadata": {},
   "source": [
    "#### Outputs Saved\r\n",
    "\r\n",
    "This notebook generated and saved a gene-level differential expression table indexed by Entrez Gene ID, representing aggregated probe-level transcriptional changes in Duchenne Muscular Dystrophy. A cleaned probe-to-gene annotation table derived from the GPL570 platform was also saved to ensure reproducibility of downstream analyses.\r\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5164f2d7-bf5c-4b1f-b9a9-a5cdd79f91ea",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
